### PR TITLE
Fix #3249: Added JWT as Auth Token

### LINKED ISSUE #3249: Backend: Submit page on EvalAI for code upload challenges shows old auth token
**Issue**: Since we recently changed to JWT auth token, the submit page for code upload challenges on EvalAI still shows the old token.
image

**Fix**:
- [x] Replace the current auth token with the  new token.

### PR DESCRIPTION
**Pull request by ShauryaAg**
Fixes #3249 

After fix:
image

### CODE REVIEW

**Comment by ShauryaAg** [comment]
@RishabhJain2018 @Ram81 I have made the necessary changes to set JWT as the Auth Token, but I still have a doubt about a piece of code.

https://github.com/Cloud-CV/EvalAI/blob/adde7be63e25af834d65ef2625103040122f561e/settings/common.py#L355

Here, you are using `RefreshToken` to prove authentication, but I think it will much more sense to use `AccessToken` instead.
Should I change this too?

**Review by RishabhJain2018 — COMMENTED** [review]
@ShauryaAg Is this PR cpmplete?

**Comment by ShauryaAg** [comment]
@RishabhJain2018 The PR is complete, and it does the necessary function of using `JWT` instead of `ExpiringAuthToken`. But please take a look at changing `RefreshToken` to `AccessToken` as I have mentioned above.

File: settings/common.py [hunk 1]
```diff
@@ -348,7 +350,7 @@
     "ALGORITHM": "HS256",
     "AUDIENCE": None,
     "ISSUER": None,
-    "AUTH_HEADER_TYPES": ("Bearer",),
```
Comment (RishabhJain2018) [inline settings/common.py:351]: Why are we changing this?

File: settings/common.py [hunk 2]
```diff
@@ -156,6 +156,8 @@
 
 SITE_ID = 1
 
+REST_USE_JWT = True
```
Comment (RishabhJain2018) [inline settings/common.py:159]: Why do we need to add this?

File: evalai/urls.py [hunk 3]
```diff
@@ -21,7 +21,7 @@
 from django.conf.urls.static import static
 
 from allauth.account.views import ConfirmEmailView
-from rest_framework_expiring_authtoken.views import obtain_expiring_auth_token
+from rest_framework_simplejwt.views import token_obtain_pair
```
Comment (RishabhJain2018) [inline evalai/urls.py:24]: Why are we changing this?

File: evalai/urls.py [hunk 4]
```diff
@@ -49,8 +49,8 @@
     url(r"^api/admin/", admin.site.urls),
     url(
         r"^api/auth/login",
-        obtain_expiring_auth_token,
-        name="obtain_expiring_auth_token",
+        token_obtain_pair,
```
Comment (RishabhJain2018) [inline evalai/urls.py:52]: Why are we changing this?

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Comment by RishabhJain2018** [comment]
> >@RishabhJain2018 @Ram81 I have made the necessary changes to set JWT as the Auth Token, but I still have a doubt about a piece of code.
> 
> EvalAI/settings/common.py
> 
> Line 355 in adde7be
> 
>  "AUTH_TOKEN_CLASSES": ("rest_framework_simplejwt.tokens.RefreshToken",), 
> Here, we are using RefreshToken to prove authentication, but I think it will make much more sense to use AccessToken instead.
> Should I change this too?

We are using `RefreshToken` because we want a way to blacklist some of the tokens. IIRC, it isn't possible with `Access Token`.
cc: @Ram81

File: settings/common.py (hunk 1 above)
Comment (ShauryaAg) [reply settings/common.py:351]: This is changed so that `rest_auth` can verify the user after login. If we don't make this change, we'll need to make changes in the frontend to set default `Authorization` headers as `Bearer <Token>` (currently, it sets default `Authorization` header as `Token <Token>`, [here](https://github.com/Cloud-CV/EvalAI/blob/adde7be63e25af834d65ef2625103040122f561e/frontend_v2/src/app/services/api.service.ts#L26))

File: settings/common.py (hunk 2 above)
Comment (ShauryaAg) [reply settings/common.py:159]: This changed is need for `rest_auth` to verify user through `JWT`.

File: evalai/urls.py (hunk 3 above)
Comment (ShauryaAg) [reply evalai/urls.py:24]: Previously, we were using `obtain_expiring_auth_token` which returns a user `Token` (**not** `JWT`), `token_obtain_pair` method returns the `Access Token JWT` and `Refresh Token JWT`.

File: evalai/urls.py (hunk 4 above)
Comment (ShauryaAg) [reply evalai/urls.py:52]: This is to replace `obtain_expiring_auth_token` with `token_obtain_pair` for login.

Comment (Ram81) [reply evalai/urls.py:52]: @ShauryaAg we can't use this. If you try login/logout multiple times the JWT token will keep changing. We are using JWT tokens mainly for API and CLI api calls. The expiring token is used for frontend codebase. For this change you just have to add an API call for `get_auth_token` [API](https://github.com/Cloud-CV/EvalAI/blob/master/apps/accounts/views.py#L43) and use the token returned by this API to show on `Submit` page.

**Comment by Ram81** [comment]
Yes, that's correct

File: settings/common.py [hunk 5]
```diff
@@ -348,7 +350,7 @@
     "ALGORITHM": "HS256",
     "AUDIENCE": None,
     "ISSUER": None,
-    "AUTH_HEADER_TYPES": ("Bearer",),
+    "AUTH_HEADER_TYPES": ("Token",),
```
Comment (Ram81) [inline settings/common.py:353]: @ShauryaAg we are not switching to `AccessToken`. We are using `RefreshToken` because we want to support token blacklisting. Can you revert this change.

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: settings/common.py (hunk 5 above)
Comment (ShauryaAg) [reply settings/common.py:353]: @Ram81 I did not change `RefreshToken` to `AccessToken`, it was only something I suggest as I was curious.

However, changing `Bearer` to `Token` is still needed, as `rest_auth` verifies and returns using request headers with `Authorization: Token <Token>` as set in the frontend.
https://github.com/Cloud-CV/EvalAI/blob/adde7be63e25af834d65ef2625103040122f561e/frontend_v2/src/app/services/api.service.ts#L26

Here, we are setting default headers for `Authorization` as `Token <Token>`. If we don't change `Bearer` to `Token` in `AUTH_HEADER_TYPES`, then we will need to set `Authorization` header as `Bearer <Token>` in the frontend.

File: evalai/urls.py (hunk 4 above)
Comment (ShauryaAg) [reply evalai/urls.py:52]: Currently, the `authtoken` is set in the local storage of the browser during login. Should I just make changes in the frontend to call that `get_auth_token` method everytime instead of setting the `authtoken` in local storage as before?

File: settings/common.py (hunk 5 above)
Comment (Ram81) [reply settings/common.py:353]: We don't want to switch to Bearer token on frontend side. To show the `RefreshToken` on the submit page just make an API call to `get_auth_token` API and use the response to show the token on UI. No need to change the authorization headers on frontend local storage

File: evalai/urls.py (hunk 4 above)
Comment (Ram81) [reply evalai/urls.py:52]: Yes, or you could make API call once and store the bearer token as a separate variable in local storage.

**Comment by ShauryaAg** [comment]
I have made the changes in Frontend as well as Frontend_V2 to set `refreshJWT` in local storage during login. Please review.

**Comment by ShauryaAg** [comment]
- Frontend after change: 
image

- Frontend_V2 after change:
image

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
Hey @ShauryaAg, On the UI, can we please show the token up to say 20 characters and then show `...` after it. It currently looks a bit ugly on UI.

Also, please make sure that when it is copied then the complete token is copied.

**Comment by ShauryaAg** [comment]
Frontend After Fix:
image

Frontend_V2 After Fix:
image

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
LGTM, code-wise.
@Ram81 Can you please check on your local once if this works?

**Comment by Ram81** [comment]
@RishabhJain2018 it works fine. Verified locally.

**Review by RishabhJain2018 — APPROVED** [review]
